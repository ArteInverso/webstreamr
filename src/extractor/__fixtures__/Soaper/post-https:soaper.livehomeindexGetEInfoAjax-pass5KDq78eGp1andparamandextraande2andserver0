{"key":true,"val":"\/home\/index\/TVM3U8?key=RnX8mnL4l5hEz38dxBAbFjV27jbe3Yh13LqxV48ViM7j60OLjmh3w5PJ8nQPh2brz0eEAAt50Zvv7d89hpM3OLWQ6nsjb4n0l3ApSna13LB3EbUxbbyJ70KwuVjaooMP5mIE4MbjyvqJHjRq3amYA5SPV8NRl8wmivl8YyrlMmFx6VomOAp2C5WR46w10v.m3u8","vtt":"https:\/\/ttt5.randchange.top\/\/hls\/2011\/black_mirror\/Season4\/2\/vtt\/\/thumb.vtt","val_bak":"\/home\/index\/TVM3U8?key=RnX8mnL4l5hEz38dxBAbFjV27jbe3Yh13LqxV48ViM7j60OLjmh3w5PJ8nQPh2brz0eEAAt50Zvv7d89hpM3OLWQ6nsjb4n0l3ApSna13LB3EbUxbbyJ70KwuVjaooMP5mIE4MbjyvqJHjRq3amYA5SPV8NRl8wmivl8YyrlMmFx6VomOAp2C5WR46w10v.m3u8","pos":0,"type":"m3u8","subs":[{"path":"\/home\/index\/getsrt?key=21ZbK1AB7YHPvy2d365Qf7A6d7Ox9jfbmPWyprvpf10rL2xor3Cndq3BJva3cP3Ovd0LEECvdPLL1r03cLX7m9BWybFqrw67eP2mH2q0E4ZE71UbrrWQaMo1tJyY226R5EhRalNQPp1dslnMX7bwpNs1LJqWbJV2TWK9EZdK3jU87lwBMprYI7R84zBEpKhXqmEYa5xQF5PbmA6KwBS86nXW4X1oh6K7y5X2O4HX8qPV76opTyB73AJPV4T215N5bRv2SrMZjKwb43HXmmJqRexPs2zZ6n2vAdfREPpP2zOyCOLnz4YqoVHyQWMyXNJMCO7q9yQZMVTQ4aMwzZ8YIAxBy7a7YbcxdA2M4vNLSjyqNA9E4Ls45jV04zvnc5AO4nNLvKsn6v9pK6jmSVdBlq3Lz4SPdKYJW.srt","name":"en"},{"path":"\/home\/index\/getsrt?key=nv0KMvQzYRC8wZz5KRELirnNMraXJRcn4VqYd02dUxW7y5K379iZ2d56W7r5c7Me612055FQ8m99weWAIM0VlyE89zU0QwN81EAVIMeWBE8Bx6Fann2BRWXzf78mwwny6oir8x5BMVJ6HeaYn3EBjwFXVLox8L36sK7x9o67mvcE15vxZzb9SqmwQ4PKZnCpyA8ojR15ilyaB6vxrRcZzvPYLPbqSJOYVvqAK4T94Z0nYNvAH1QJ8LYnvzSlorwr4KylFb5aAp3LQYiBRRn417VpSdY76Mdl9WuylEzEKjPYCMXZ0V9qexh0xBW0N3wWSK9p51wJVviA4yN3w0m2Iq01r2d2WQInYajrJbovIpoB4EK38eip1Je6pKvxubN53JZwE7h9YzyjOYdAULxabZbYVRCnbAnr5Mvqip1q5WO.srt","name":"en:hi"},{"path":"\/home\/index\/getsrt?key=Lre35r89YAh4O1KRQe0duLdBzLNyRaf8zPmE6qo6S1Mrj60yrmCvAoKOqY2KT5Kxp09ennUz20YYPQ1yu3wROMp4LyTYRA4Xyq0Ziw4rAnJA3ahnddr1mKyOF6Ja77xLmyuJ5ZMpoxrnfBlonvEWrzS1YXqMwXd6T7LbVjqLRwu0rVoWvljzhpZ5q6bvPdTALJaV0rbQHNxRBr316zf1pKN75Nq2i7MoEdqjZRi024moe1yNcMdmrq4xjptnA3P3O7enT1xQvYzyBphynnBJeOAluBvpONBbz3IQNR4RjZ6Wu9QqV1dKoyfeNVjeoqnjCdAyMbRLnrUwvYX14eKMtj0rZ2b2APIMjAd2OJrotXVAJ0LNE1Hpwn7Lp0O6ulYVn6x0bQtbe6d5MPM7CB7QqwKqWmIjz2X2lnQKFleErlJB6oSbNWv908YqinMZ1JM4KmtO79pOXZR6FYePVNNoyNHLV4Q0o93EI5nrP1KvLNFK0Qj66qOOh7Aw46LP5Ps1nQoAyELoFme4MzM53dUzy9.srt","name":"Encoded By nItRo And Uploaded By XpoZ.srt"}],"prev_epi_title":"[S4E1] USS Callister","prev_epi_url":"\/episode_4vGPvjBD31.html?ap=1","next_epi_title":"[S4E3] Crocodile","next_epi_url":"\/episode_PnG6nAnG7v.html?ap=1"}