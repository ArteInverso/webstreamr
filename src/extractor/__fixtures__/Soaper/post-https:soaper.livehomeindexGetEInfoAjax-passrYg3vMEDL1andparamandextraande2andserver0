{"key":true,"val":"\/home\/index\/TVM3U8?key=LrezaYpLOxT4O61m8e3mSLdBzLNyRaf8zPmERxvwh1Mrj6zbVEUvAzRv2KZLu5PVvRKqwbcz6qmb9vbOi3wQZ4ae8RsZ9N3Jm9Yatw4yE55bBp.m3u8","vtt":"https:\/\/ttt5.randchange.top\/\/hls\/2023\/last_of_us\/Season2\/3\/vtt\/\/thumb.vtt","val_bak":"\/home\/index\/TVM3U8?key=LrezaYpLOxT4O61m8e3mSLdBzLNyRaf8zPmERxvwh1Mrj6zbVEUvAzRv2KZLu5PVvRKqwbcz6qmb9vbOi3wQZ4ae8RsZ9N3Jm9Yatw4yE55bBp.m3u8","pos":0,"type":"m3u8","subs":null,"prev_epi_title":"[S2E2] Through the Valley","prev_epi_url":"\/episode_dZkrY8nG6z.html?ap=1","next_epi_title":"[S2E4] Day One","next_epi_url":"\/episode_OMkKZ9WGLq.html?ap=1"}