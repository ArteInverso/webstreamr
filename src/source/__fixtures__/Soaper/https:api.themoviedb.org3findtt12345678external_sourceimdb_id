{"movie_results":[{"backdrop_path":null,"id":1427398,"title":"Hold & Kick","original_title":"Hold & Kick","overview":"","poster_path":null,"media_type":"movie","adult":false,"original_language":"id","genre_ids":[],"popularity":0.0071,"release_date":"","video":false,"vote_average":0.0,"vote_count":0}],"person_results":[],"tv_results":[],"tv_episode_results":[],"tv_season_results":[]}