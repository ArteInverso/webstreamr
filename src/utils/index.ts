export * from './env';
export * from './Fetcher';
export * from './StreamResolver';
export * from './embed';
export * from './imdb';
export * from './languageFromCountryCode';
export * from './manifest';
export * from './tmdb';
