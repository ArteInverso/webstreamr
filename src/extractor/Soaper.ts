import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta, UrlResult } from '../types';

interface SoaperInfoResponsePartial {
  val: string;
  val_bak: string;
}

export class Soaper implements Extractor {
  readonly id = 'soaper';

  readonly label = 'Soaper';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/soaper/) && null !== url.pathname.match(/^\/(episode|movie)_/);

  readonly normalize = (url: URL): URL => url;

  readonly extract = async (ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> => {
    const movieOrEpisodeId = (url.pathname.match(/\/\w+_(\w+)/) as string[])[1] as string;

    const form = new URLSearchParams();
    form.append('pass', movieOrEpisodeId);
    form.append('param', '');
    form.append('extra', '');
    form.append('e2', '');
    form.append('server', '0');

    const response = await this.fetcher.textPost(
      ctx,
      new URL(url.pathname.includes('episode') ? '/home/index/GetEInfoAjax' : '/home/index/GetMInfoAjax', url.origin),
      form.toString(),
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );
    const jsonResponse = JSON.parse(response) as SoaperInfoResponsePartial;

    const m3u8Url = new URL(jsonResponse['val'], url.origin);
    const m3u8Data = await this.fetcher.text(ctx, m3u8Url);
    const height = m3u8Data.match(/\d+x(\d+)|(\d+)p/) as string[];

    return [
      {
        url: new URL(jsonResponse['val'], url.origin),
        label: this.label,
        sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          ...(height && { height: parseInt(height[1] ?? height[2] as string) }),
        },
      },
    ];
  };
}
