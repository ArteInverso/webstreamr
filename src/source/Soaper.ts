import { ContentType } from 'stremio-addon-sdk';
import * as cheerio from 'cheerio';
import { Source, SourceResult } from './types';
import { Fetcher, getTmdbId, getTmdbMovieDetails, getTmdbTvDetails, Id, TmdbId } from '../utils';
import { Context, CountryCode } from '../types';

export class Soaper implements Source {
  public readonly id = 'soaper';

  public readonly label = 'Soaper';

  public readonly contentTypes: ContentType[] = ['movie', 'series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.en];

  private readonly baseUrl = 'https://soaper.live';

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public async handle(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const [keyword, year, hrefPrefix] = await this.getKeywordYearAndHrefPrefix(ctx, tmdbId);

    const pageUrl = await this.fetchPageUrl(ctx, keyword, year, hrefPrefix);
    if (!pageUrl) {
      return [];
    }

    if (tmdbId.season) {
      const html = await this.fetcher.text(ctx, pageUrl);

      const $ = cheerio.load(html);

      const episodePageHref = $(`.alert:has(h4:contains("Season${tmdbId.season}")) a:contains("${tmdbId.episode}.")`).attr('href');
      if (!episodePageHref) {
        return [];
      }

      const episodeUrl = new URL(episodePageHref, this.baseUrl);
      const title = `${keyword} ${tmdbId.season}x${tmdbId.episode}`;

      return [{ countryCode: CountryCode.en, title, url: episodeUrl }];
    }

    const title = `${keyword} (${year})`;

    return [{ countryCode: CountryCode.en, title, url: pageUrl }];
  };

  private readonly fetchPageUrl = async (ctx: Context, keyword: string, year: number, hrefPrefix: string): Promise<URL | undefined> => {
    const searchUrl = new URL(`/search.html?keyword=${encodeURIComponent(keyword)}`, this.baseUrl);
    const html = await this.fetcher.text(ctx, searchUrl);

    const $ = cheerio.load(html);

    return $(`.thumbnail .img-group:has(.img-tip:contains("${year}")) a[href^="${hrefPrefix}"]`)
      .map((_i, el) => new URL($(el).attr('href') as string, this.baseUrl))
      .get(0);
  };

  private readonly getKeywordYearAndHrefPrefix = async (ctx: Context, tmdbId: TmdbId): Promise<[string, number, string]> => {
    if (tmdbId.season) {
      const tmdbDetails = await getTmdbTvDetails(ctx, this.fetcher, tmdbId);

      return [tmdbDetails.name, (new Date(tmdbDetails.first_air_date)).getFullYear(), '/tv_'];
    }

    const tmdbDetails = await getTmdbMovieDetails(ctx, this.fetcher, tmdbId);

    return [tmdbDetails.title, (new Date(tmdbDetails.release_date)).getFullYear(), '/movie_'];
  };
}
