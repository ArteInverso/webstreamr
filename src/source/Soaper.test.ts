import { createTestContext } from '../test';
import { FetcherMock, ImdbId, TmdbId } from '../utils';
import { Soaper } from './Soaper';

const ctx = createTestContext();

describe('Soaper', () => {
  let handler: Soaper;

  beforeEach(() => {
    handler = new Soaper(new FetcherMock(`${__dirname}/__fixtures__/Soaper`));
  });

  test('handles non-existent movies gracefully', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt12345678', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle non-uploaded movie gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt29141112', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle non-existent episode gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt2085059', 99, 99));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb black mirror s4e2', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt2085059', 4, 2));
    expect(streams).toMatchSnapshot();
  });

  test('handle tmdb black mirror s4e2', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(42009, 4, 2));
    expect(streams).toMatchSnapshot();
  });

  test('handle imdb full metal jacket', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt0093058', undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
