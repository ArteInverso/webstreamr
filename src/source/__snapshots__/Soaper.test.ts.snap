// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Soaper handle imdb black mirror s4e2 1`] = `
[
  {
    "countryCode": "en",
    "title": "Black Mirror 4x2",
    "url": "https://soaper.live/episode_5KDq78eGp1.html",
  },
]
`;

exports[`Soaper handle imdb full metal jacket 1`] = `
[
  {
    "countryCode": "en",
    "title": "Full Metal Jacket (1987)",
    "url": "https://soaper.live/movie_d8kdeypDY9.html",
  },
]
`;

exports[`Soaper handle tmdb black mirror s4e2 1`] = `
[
  {
    "countryCode": "en",
    "title": "Black Mirror 4x2",
    "url": "https://soaper.live/episode_5KDq78eGp1.html",
  },
]
`;
